session.go

```diff
@@ -384,21 +385,66 @@ runLoop:
 		}
 
 		now := time.Now()
+
 		if timeout := s.sentPacketHandler.GetAlarmTimeout(); !timeout.IsZero() && timeout.Before(now) {
 			// This could cause packets to be retransmitted, so check it before trying
 			// to send packets.
 			s.sentPacketHandler.OnAlarm()
 		}
 
-		if s.config.KeepAlive && s.handshakeComplete && time.Since(s.lastNetworkActivityTime) >= s.peerParams.IdleTimeout/2 {
+		if s.config.KeepAlive && !s.keepAlivePingSent && s.handshakeComplete && time.Since(s.lastNetworkActivityTime) >= s.peerParams.IdleTimeout/2 {
 			// send the PING frame since there is no activity in the session
 			s.packer.QueueControlFrame(&wire.PingFrame{})
 			s.keepAlivePingSent = true
+		} else if !pacingDeadline.IsZero() && now.Before(pacingDeadline) {
+			// If we get to this point before the pacing deadline, we should wait until that deadline.
+			// This can happen when scheduleSending is called, or a packet is received.
+			// Set the timer and restart the run loop.
+			s.maybeResetTimer(pacingDeadline)
+			continue
+		}
+
+		var startPacingTimer bool
+		numPackets, pacingDelay := s.sentPacketHandler.TimeUntilSend(now)
+		packetsSent := 1
+		if !s.sentPacketHandler.SendingAllowed() { // if congestion limited, at least try sending an ACK frame
+			if err := s.maybeSendAckOnlyPacket(); err != nil {
```

Comment:
I don't understand why sending an ACK-only packet doesn't have an effect on pacing. It can change state in the SPH, so you need similar logic to the normal case, right?

Reply:
If we're congestion limited, we don't care about pacing (since pacing is not the limiting factor anyway).